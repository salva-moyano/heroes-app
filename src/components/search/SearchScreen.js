import React, {useMemo} from "react";
import {useForm} from "../../hooks/UseForm";
import {getHeroesByName} from "../../selectors/getHeroesByName";
import {HeroCard} from "../hero/HeroCard";
import {useNavigate, useLocation} from "react-router-dom";
import queryString from 'query-string';

export const SearchScreen = () => {

    const navigate = useNavigate();
    const location = useLocation();
    const {q = ''} = queryString.parse(location.search);

    const [formValues, handleInputChange, reset]  = useForm({
        searchText: q
    })
    const { searchText } = formValues;

    const heroesFiltered =  useMemo(() => getHeroesByName(q), [q])

    const handleSearch = (e) => {
        e.preventDefault();
        navigate(`?q=${searchText}`)
    }

    return(
        <>
            <div className="row">
                <div className="col-5">
                    <h4>Buscar</h4>
                    <hr/>
                    <form onSubmit={ handleSearch }>
                        <input type="text"
                               placeholder="Buscar un héroe"
                               className="form-control"
                               autoComplete="off"
                               value={ searchText }
                               onChange={ handleInputChange }
                               name="searchText"/>
                        <div className="d-grid gap-2">
                            <button type="submit"
                                className="btn btn-outline-primary mt-3">
                                Buscar...
                            </button>
                        </div>
                    </form>
                </div>
                <div className="col-7">
                    <h4>Resultados</h4>

                    {
                        (q === '') ? <div className="alert alert-info animate__animated animate__fadeIn">Buscar un héroe</div>
                            : ( heroesFiltered.length === 0) && <div className="alert alert-danger animate__animated animate__fadeIn">No hay resultados {q}</div>
                    }

                    <hr />
                    {
                      heroesFiltered.map(hero => (
                            <HeroCard key={hero.id} {...hero} />
                      ))
                    }
                </div>
            </div>
        </>
    )
}