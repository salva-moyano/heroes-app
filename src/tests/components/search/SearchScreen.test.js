import {mount } from 'enzyme';

import {SearchScreen} from "../../../components/search/SearchScreen";
import {MemoryRouter} from "react-router-dom";

const mockNavigate = jest.fn();

jest.mock("react-router-dom", () => (
    {
        ...jest.requireActual('react-router-dom'),
        useNavigate: () => mockNavigate
    }));

describe('Testing <SearchScreen', () => {

    test('should show successfully with default values', () => {

        const wrapper = mount(
            <MemoryRouter initialEntries={['/search']}>
                <SearchScreen/>
            </MemoryRouter>
        )

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('.alert-info').text().trim()).toBe('Buscar un hÃ©roe');

    });

    test('should show a Batman hero and input with value Batman', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <SearchScreen/>
            </MemoryRouter>
        )

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('input').prop('value')).toBe('batman')
        expect(wrapper.find('h5').text().trim()).toBe('Batman')
    });

    test('should show a error if not found hero', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=qwererere']}>
                <SearchScreen/>
            </MemoryRouter>
        )

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('.alert-danger').text().trim()).toBe('No hay resultados qwererere');
    });

    test('should call the navigate new screen', () => {

        const wrapper = mount(
            <MemoryRouter initialEntries={['/search']}>
                <SearchScreen/>
            </MemoryRouter>
        );

        wrapper.find('input').simulate('change', {target: {name: 'searchText', value: 'batman'}});
        wrapper.find('form').prop('onSubmit')({ preventDefault() {} })
        expect(mockNavigate).toHaveBeenCalled()
        expect(mockNavigate).toHaveBeenCalledWith('?q=batman')
    });

});