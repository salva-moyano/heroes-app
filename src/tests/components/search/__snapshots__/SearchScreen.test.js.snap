// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing <SearchScreen should show a Batman hero and input with value Batman 1`] = `
<div
  className="row"
>
  <div
    className="col-5"
  >
    <h4>
      Buscar
    </h4>
    <hr />
    <form
      onSubmit={[Function]}
    >
      <input
        autoComplete="off"
        className="form-control"
        name="searchText"
        onChange={[Function]}
        placeholder="Buscar un héroe"
        type="text"
        value="batman"
      />
      <div
        className="d-grid gap-2"
      >
        <button
          className="btn btn-outline-primary mt-3"
          type="submit"
        >
          Buscar...
        </button>
      </div>
    </form>
  </div>
  <div
    className="col-7"
  >
    <h4>
      Resultados
    </h4>
    <hr />
    <div
      className="card animate__animated animate__fadeIn"
    >
      <div
        className="row no-gutters"
      >
        <div
          className="col-6"
        >
          <img
            alt="Batman"
            className="card-img"
            src="../assets/heroes/dc-batman.jpg"
          />
        </div>
        <div
          className="col-6"
        >
          <div
            className="card-body"
          >
            <h5
              className="card-title"
            >
              Batman
            </h5>
            <p
              className="card-text"
            >
              Bruce Wayne
            </p>
            <p
              className="card-text"
            >
              <small
                className="text-muted"
              >
                 
                Detective Comics #27
              </small>
            </p>
            <a
              href="/hero/dc-batman"
              onClick={[Function]}
            >
              Más...
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Testing <SearchScreen should show a error if not found hero 1`] = `
<div
  className="row"
>
  <div
    className="col-5"
  >
    <h4>
      Buscar
    </h4>
    <hr />
    <form
      onSubmit={[Function]}
    >
      <input
        autoComplete="off"
        className="form-control"
        name="searchText"
        onChange={[Function]}
        placeholder="Buscar un héroe"
        type="text"
        value="qwererere"
      />
      <div
        className="d-grid gap-2"
      >
        <button
          className="btn btn-outline-primary mt-3"
          type="submit"
        >
          Buscar...
        </button>
      </div>
    </form>
  </div>
  <div
    className="col-7"
  >
    <h4>
      Resultados
    </h4>
    <div
      className="alert alert-danger animate__animated animate__fadeIn"
    >
      No hay resultados 
      qwererere
    </div>
    <hr />
  </div>
</div>
`;

exports[`Testing <SearchScreen should show successfully with default values 1`] = `
<div
  className="row"
>
  <div
    className="col-5"
  >
    <h4>
      Buscar
    </h4>
    <hr />
    <form
      onSubmit={[Function]}
    >
      <input
        autoComplete="off"
        className="form-control"
        name="searchText"
        onChange={[Function]}
        placeholder="Buscar un héroe"
        type="text"
        value=""
      />
      <div
        className="d-grid gap-2"
      >
        <button
          className="btn btn-outline-primary mt-3"
          type="submit"
        >
          Buscar...
        </button>
      </div>
    </form>
  </div>
  <div
    className="col-7"
  >
    <h4>
      Resultados
    </h4>
    <div
      className="alert alert-info animate__animated animate__fadeIn"
    >
      Buscar un héroe
    </div>
    <hr />
  </div>
</div>
`;
